linker/editions.go

```diff
@@ -0,0 +1,167 @@
+)
+
+type hasFeatures interface {
+	GetFeatures() *descriptorpb.FeatureSet
+}
+
+var _ hasFeatures = (*descriptorpb.FileOptions)(nil)
+var _ hasFeatures = (*descriptorpb.MessageOptions)(nil)
+var _ hasFeatures = (*descriptorpb.FieldOptions)(nil)
+var _ hasFeatures = (*descriptorpb.OneofOptions)(nil)
+var _ hasFeatures = (*descriptorpb.ExtensionRangeOptions)(nil)
+var _ hasFeatures = (*descriptorpb.EnumOptions)(nil)
+var _ hasFeatures = (*descriptorpb.EnumValueOptions)(nil)
+var _ hasFeatures = (*descriptorpb.ServiceOptions)(nil)
+var _ hasFeatures = (*descriptorpb.MethodOptions)(nil)
+
+// resolveFeature resolves a feature for the given descriptor. It uses the function
+// to extract a value from a feature set. The function should return false as the
+// second argument if the first value it returns was not actually present in the
+// feature set.
+func resolveFeature[T any](d protoreflect.Descriptor, accessor func(*descriptorpb.FeatureSet) (T, bool)) T {
+	if d == nil {
+		var zero T
+		return zero
+	}
+	for {
+		var features *descriptorpb.FeatureSet
+		if withFeatures, ok := d.Options().(hasFeatures); ok {
+			// It should not be possible for 'ok' to ever be false...
+			features = withFeatures.GetFeatures()
+		}
+		val, ok := accessor(features)
+		if ok {
+			return val
+		}
+		parent := d.Parent()
+		if parent == nil {
+			// We've reached the end of the inheritance chain. So we fall back to a default.
+			return getFeatureDefault(d, accessor)
+		}
+		d = parent
+	}
+}
+
+func getFeatureDefault[T any](d protoreflect.Descriptor, accessor func(*descriptorpb.FeatureSet) (T, bool)) T {
+	defaults := getEditionDefaults(getEdition(d))
+	val, _ := accessor(defaults)
+	return val
+}
+
+type hasEdition interface {
+	// Edition returns the numeric value of a google.protobuf.Edition enum
+	// value that corresponds to the edition of this file. If the file does
+	// not use editions, it should return the enum value that corresponds
+	// to the syntax level, EDITION_PROTO2 or EDITION_PROTO3.
+	Edition() int32
+}
+
+var _ hasEdition = (*result)(nil)
+
+func getEdition(d protoreflect.Descriptor) descriptorpb.Edition {
+	withEdition, ok := d.ParentFile().(hasEdition)
+	if !ok {
+		// The parent file should always be a *result, so we should
+		// never be able to actually get in here. If we somehow did
+		// have another implementation of protoreflect.FileDescriptor,
+		// it doesn't provide a way to get the edition, other than the
+		// potentially expensive step of generating a FileDescriptorProto
+		// and then querying for the edition from that. :/
+		return descriptorpb.Edition_EDITION_UNKNOWN
+	}
+	return descriptorpb.Edition(withEdition.Edition())
+}
+
+func getEditionDefaults(edition descriptorpb.Edition) *descriptorpb.FeatureSet {
+	editionDefaultsInit.Do(func() {
+		editionDefaults = make(map[descriptorpb.Edition]*descriptorpb.FeatureSet, len(descriptorpb.Edition_name))
+		// Compute default for all known editions in descriptorpb.
```

Comment:
This compiles feature-set defaults from descriptor.proto compiled into `descriptorpb`, and does not look at a potential google/protobuf/descriptor.proto fed to the compiler, correct?

Since this is only used to construct `protoreflect` types, I can't think of a case where this wouldn't be sufficient 🤷

Reply:
That is correct. I am merely trying to implement these interfaces consistently with how they are implemented elsewhere in the protobuf-go runtime. So if you compiled files with a custom descriptor.proto and generated Go code, that Go code would act this same way, since its descriptor implementations have the same version of descriptor.proto "baked in".